scripts/run_frontend_tests.py

```diff
@@ -130,9 +170,50 @@ def main(args: Optional[Sequence[str]] = None) -> None:
         'Running test in development environment'])
 
     cmd = [
-            common.NODE_BIN_PATH, '--max-old-space-size=4096',
-            os.path.join(common.NODE_MODULES_PATH, 'karma', 'bin', 'karma'),
-            'start', os.path.join('core', 'tests', 'karma.conf.ts')]
+        common.NODE_BIN_PATH, '--max-old-space-size=4096',
+        os.path.join(common.NODE_MODULES_PATH, 'karma', 'bin', 'karma'),
+        'start', os.path.join('core', 'tests', 'karma.conf.ts')]
+
+    specs_to_run: Set[str] = set()
+    if parsed_args.specs_to_run:
+        for spec in parsed_args.specs_to_run.split(','):
+            spec_file = get_file_spec(spec.strip())
+            if spec_file:
+                specs_to_run.add(spec_file)
+
+    if parsed_args.run_on_changed_files:
+        remote = git_changes_utils.get_local_git_repository_remote_name()
+        if not remote:
+            sys.exit('Error: No remote repository found.')
+        refs = git_changes_utils.get_refs()
+        collected_files = git_changes_utils.get_changed_files(
+            refs, remote.decode('utf-8'))
+        for _, (_, acmrt_files) in collected_files.items():
+            if not acmrt_files:
+                continue
+
+            files_to_run = git_changes_utils.get_js_or_ts_files_from_diff(
+                acmrt_files)
+            for file_path in files_to_run:
+                spec_file = get_file_spec(file_path)
+                if spec_file:
+                    specs_to_run.add(spec_file)
+        staged_files = git_changes_utils.get_staged_files()
+        staged_files_to_run = git_changes_utils.get_js_or_ts_files_from_diff(
+            staged_files)
+        for file_path in staged_files_to_run:
+            spec_file = get_file_spec(file_path)
+            if spec_file:
+                specs_to_run.add(spec_file)
+
+    if parsed_args.specs_to_run or parsed_args.run_on_changed_files:
+        if len(specs_to_run) == 0:
+            print('No valid specs found to run.')
+            sys.exit(0)
```

Comment:
This should be a failure condition so that it results in a CI failure

Reply:
Done.

Reply:
Actually regarding this, I wanted to exit with code 0 since I didn't want the pre-push hook to fail (as it just passes all ts/js changed files to the frontend check script), and this isn't used in the CI at all.

Reply:
Hmm yes I see that this doesn't get used on CI, but why don't you want the pre-push hook to fail if the user specified specs to run yet none were found? In that situation it seems very likely that something's wrong.

On the other hand, I think it's reasonable for the pre-push hook to pass if the user only asked to run tests on changed files and there were no specs to run. That's totally possible if the user just didn't change any frontend files. You might need to handle these two cases separately

Reply:
The pre-push hook just gets the changed ts or js files, this doesn't necessarily mean that it has a spec file (combined spec.ts .lighthouserc.js, etc.), but they are passed to the frontend test script anyway and that script will detect if there are any specs to run or not.

Reply:
Ahh sorry, I mis-phrased my earlier reply. Here is what I meant to say:

> Hmm yes I see that this doesn't get used on CI, but why don't you want the `run_frontend_tests.py` script to fail if the user specified specs to run yet none were found? In that situation it seems very likely that something's wrong.
>
> On the other hand, I think it's reasonable for the pre-push hook to pass if the user only asked to run tests on changed files and there were no specs to run. That's totally possible if the user just didn't change any frontend files. You might need to handle these two cases separately

To be more explicit, this `if` block will run if the user manually runs the frontend tests and passes specific spec files to run using `--specs_to_run`. In that case, if no spec files were found, it seems wrong to return a status code of 0 since the user was clearly expecting some spec files to be executed, right?

Reply:
That would work. Alternatively, since we're carving out special behavior for the pre-push hook, you could also just add a flag like `--allow-no-spec` that causes the script to exit with code 0 even if there are no spec files found. By default, we get the safer behavior of returning a code of 1, but the pre-push-hook can pass `--allow-no-spec` to get our desired behavior. This approach might be simpler to implement

Reply:
Done added a flag, Thanks.